File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/HomeFeedCache.kt
package com.bockmedia.console.domain.model

/** In-memory home feed cache so tab switches and pull-refresh feel instant. */
object HomeFeedCache {
    private const val TTL_MS = 600_000L // 10 min — tab switches stay instant

    @Volatile
    var feed: HomeFeed? = null
        private set

    @Volatile
    private var cachedAtMs: Long = 0L

    @Volatile
    private var hasRatedSongs: Boolean? = null

    /** Last cached feed regardless of age — for instant tab revisit paint. */
    fun peek(): HomeFeed? = feed

    fun peekHasRatedSongs(): Boolean? = hasRatedSongs

    fun getIfFresh(): HomeFeed? {
        val cached = feed ?: return null
        if (System.currentTimeMillis() - cachedAtMs > TTL_MS) return null
        return cached
    }

    fun put(value: HomeFeed, hasRatedSongs: Boolean? = null) {
        if (value.sections.isEmpty()) return
        feed = value
        cachedAtMs = System.currentTimeMillis()
        if (hasRatedSongs != null) this.hasRatedSongs = hasRatedSongs
    }

    fun setHasRatedSongs(value: Boolean) {
        hasRatedSongs = value
    }

    fun invalidate() {
        feed = null
        cachedAtMs = 0L
        hasRatedSongs = null
    }
}

File: ios/BockMedia/Domain/HomeFeedCache.swift
import Foundation

enum HomeFeedCache {
    private static let ttlMs: Int64 = 90_000
    private static var feed: HomeFeed?
    private static var cachedAtMs: Int64 = 0
    private static var hasRatedSongs: Bool?

    static func getIfFresh() -> HomeFeed? {
        guard let cached = feed else { return nil }
        if Int64(Date().timeIntervalSince1970 * 1000) - cachedAtMs > ttlMs { return nil }
        return cached
    }

    static func peek() -> HomeFeed? { feed }

    static func peekHasRatedSongs() -> Bool? { hasRatedSongs }

    static func put(_ value: HomeFeed, hasRatedSongs rated: Bool? = nil) {
        guard !value.sections.isEmpty else { return }
        feed = value
        cachedAtMs = Int64(Date().timeIntervalSince1970 * 1000)
        if let rated { hasRatedSongs = rated }
    }

    static func setHasRatedSongs(_ value: Bool) {
        hasRatedSongs = value
    }

    static func invalidate() {
        feed = nil
        cachedAtMs = 0
        hasRatedSongs = nil
    }
}
